assets/model_monitoring/components/src/shared_utilities/io_utils.py

```diff
@@ -14,45 +34,76 @@ def init_spark():
     return spark
 
 
-def _is_not_found_exception(error: Exception):
-    return (
-        isinstance(error, IndexError)
-        or "The requested stream was not found" in error.args[0]
-        or "Not able to find MLTable file" in error.args[0]
-    )
+def _get_input_not_found_category(error: Exception):
+    if isinstance(error, IndexError) or ("Partition 0 is out of bounds." in error.args[0]):
+        return InputNotFoundCategory.NO_INPUT_IN_WINDOW
+    elif "The requested stream was not found" in error.args[0]:
+        return InputNotFoundCategory.ROOT_FOLDER_NOT_FOUND
+    elif "Not able to find MLTable file" in error.args[0]:
+        return InputNotFoundCategory.MLTABLE_NOT_FOUND
+    else:
+        return InputNotFoundCategory.NOT_INPUT_MISSING
 
 
-def try_read_mltable_in_spark_with_warning(
-    mltable_path: str, input_name: str
-) -> DataFrame:
+def try_read_mltable_in_spark_with_warning(mltable_path: str, input_name: str) -> DataFrame:
     """Read mltable in spark. In case of failure, posts a warning to the job and returns None."""
-    try:
-        return read_mltable_in_spark(mltable_path)
-    except Exception as error:
-        if _is_not_found_exception(error):
-            print(error)
-            error_message = f"No data was found for input '{input_name}'."
-            print(error_message)
+    return try_read_mltable_in_spark(mltable_path, input_name, NoDataApproach.WARNING)
+
+
+def try_read_mltable_in_spark_with_error(mltable_path: str, input_name: str) -> DataFrame:
+    """Read mltable in spark."""
+    return try_read_mltable_in_spark(mltable_path, input_name, NoDataApproach.ERROR)
+
+
+def try_read_mltable_in_spark(mltable_path: str, input_name: str, no_data_approach=NoDataApproach.IGNORE) -> DataFrame:
+    """
+    Read mltable in spark.
+
+    If data not found, conduct different error handling based on no_data_approach.
+    """
+    def process_input_not_found(input_not_found_category: InputNotFoundCategory):
+        err_msg_map = {
+            InputNotFoundCategory.NOT_INPUT_MISSING:
+                "Not input missing error.",
+            InputNotFoundCategory.NO_INPUT_IN_WINDOW: (
+                f"No data is found for input '{input_name}' in the specified window, "
+                "most likely there is no request in the specified window. "
+                f"You can check the filter field of MLTable in {mltable_path} for the time window."
+            ),
+            InputNotFoundCategory.ROOT_FOLDER_NOT_FOUND: (
```

Comment:
This is the same, is it possible to send back the path that we are trying to read

Reply:
again, the path info is in spark mltable utilities which is out of MoMo code base, and I've included mltable_path in the err msg and highlighted pattern field for the customer to check.